core/src/main/java/com/google/cloud/sql/core/CloudSqlConnectorRegistry.java

```diff
@@ -0,0 +1,361 @@
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.logging.Logger;
+import javax.net.ssl.SSLSocket;
+import jnr.unixsocket.UnixSocketAddress;
+import jnr.unixsocket.UnixSocketChannel;
+
+/**
+ * Factory responsible for obtaining an ephemeral certificate, if necessary, and establishing a
+ * secure connecting to a Cloud SQL instance.
+ *
+ * <p>This class should not be used directly, but only through the JDBC driver specific {@code
+ * SocketFactory} implementations.
+ *
+ * <p>The API of this class is subject to change without notice.
+ */
+public final class CloudSqlConnectorRegistry {
+  static final long DEFAULT_MAX_REFRESH_MS = 30000;
+  private static final Logger logger = Logger.getLogger(CloudSqlConnectorRegistry.class.getName());
+
+  private static final int DEFAULT_SERVER_PROXY_PORT = 3307;
+  private static final int RSA_KEY_SIZE = 2048;
+  private static final List<String> userAgents = new ArrayList<>();
+  private static final String version = getVersion();
+  private static final long MIN_REFRESH_DELAY_MS = 30000; // Minimum 30 seconds between refresh.
+  private static CloudSqlConnectorRegistry cloudSqlConnectorRegistry;
+  private final ListenableFuture<KeyPair> localKeyPair;
+  private final ConcurrentHashMap<String, DefaultConnectionInfoCache> connectionInfoCaches =
+      new ConcurrentHashMap<>();
+  private final ListeningScheduledExecutorService executor;
+  private final CredentialFactory credentialFactory;
+  private final int serverProxyPort;
+  private final long refreshTimeoutMs;
+  private final ConnectionInfoRepositoryFactory connectionInfoRepositoryFactory;
+
+  /**
+   * Property used to set the application name for the underlying SQLAdmin client.
+   *
+   * @deprecated Use {@link #setApplicationName(String)} to set the application name
+   *     programmatically.
+   */
+  static final String USER_TOKEN_PROPERTY_NAME = "_CLOUD_SQL_USER_TOKEN";
+
+  @VisibleForTesting
+  CloudSqlConnectorRegistry(
+      ListenableFuture<KeyPair> localKeyPair,
+      ConnectionInfoRepositoryFactory connectionInfoRepositoryFactory,
+      CredentialFactory credentialFactory,
+      int serverProxyPort,
+      long refreshTimeoutMs,
+      ListeningScheduledExecutorService executor) {
+    this.connectionInfoRepositoryFactory = connectionInfoRepositoryFactory;
+    this.credentialFactory = credentialFactory;
+    this.serverProxyPort = serverProxyPort;
+    this.executor = executor;
+    this.localKeyPair = localKeyPair;
+    this.refreshTimeoutMs = refreshTimeoutMs;
+  }
+
+  /** Returns the {@link CloudSqlConnectorRegistry} singleton. */
+  public static synchronized CloudSqlConnectorRegistry getInstance() {
```

Comment:
Does this need to be public? I know we're going to turn this into an enum as we have in AlloyDB, but shall we keep the API surface as small as possible during that refactor?

Reply:
This needs to be public so that it can be accessed by `com.google.cloud.sql.ConnectorRegistry`, which is in a different package. We can look into restricting external access to the `core` package using Java Modules some day.

Reply:
OK that's fine. Let's make it clear that this is a "private" method in that case with a similar warning.